airbyte-cdk/python/airbyte_cdk/utils/concurrent.py

```diff
@@ -0,0 +1,144 @@
+        self.delay: float = self.DELAY
+
+    def __enter__(self):
+        for _ in range(self.max_workers):
+            threading.Thread(target=self.worker, daemon=True).start()
+        threading.Thread(target=self.producer, daemon=True).start()
+        return self
+
+    def __exit__(self, exc, value, tb):
+        self.stop_all.set()
+
+    def producer(self):
+        try:
+            slices = self.stream_instance.stream_slices(sync_mode=SyncMode.full_refresh, cursor_field=self.configured_stream.cursor_field)
+            for slice_number, slice in enumerate(slices, 1):
+                q_message = QueueMessage(slice_number, 0, slice)
+                try:
+                    self.producer_to_worker.put(q_message, timeout=10)
+                except queue.Full:
+                    pass
+                if self.stop_all.is_set():
+                    return
+
+            for _ in range(self.max_workers):
+                try:
+                    self.producer_to_worker.put(messageDone, timeout=10)
+                except queue.Full:
+                    pass
+                if self.stop_all.is_set():
+                    return
+        except Exception as e:
+            self.logger.exception(e)
+            self.worker_to_consumer.put(messageFail)
+
+    def worker(self):
+        try:
+            while True:
+                if self.stop_all.is_set():
+                    return
+
+                try:
+                    item = self.producer_to_worker.get(timeout=10)
+                except queue.Empty:
+                    continue
+
+                if item == messageDone:
+                    self.worker_to_consumer.put(messageDone)
+                    break
+
+                slice = item.record
+                records = self.stream_instance.read_records(
+                    stream_slice=slice, sync_mode=SyncMode.full_refresh, cursor_field=self.configured_stream.cursor_field
+                )
+                for record_number, record in enumerate(records, 1):
+                    q_message = QueueMessage(item.slice_number, record_number, record)
+                    self.worker_to_consumer.put(q_message)
+                    if self.stop_all.is_set():
+                        return
+                q_message = QueueMessage(item.slice_number, record_number + 1)
+                self.worker_to_consumer.put(q_message)
+
+        except Exception as e:
+            self.logger.exception(e)
+            self.worker_to_consumer.put(messageFail)
+
+    def __iter__(self):
+        stopped_workers = 0
+        while True:
+            item = self.worker_to_consumer.get(timeout=300)
+            if item == messageDone:
+                stopped_workers += 1
+                if stopped_workers >= self.max_workers:
+                    return
+            elif item == messageFail:
+                return
+            else:
+                if not self.is_next(item):
+                    self.worker_to_consumer.put(item)
```

Comment:
please take a look on this part: same function uses get and put functions of same queue and it is an anti pattern.